### PR TITLE
[Feature KAN-5]added subscriptions

### PR DESCRIPTION
**Pull request by arj9912**

### CODE REVIEW

File: subscription/views.py [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+# Create your views here.
+
+
+class PlanViewSet(BaseApiMixin, ModelViewSet):
+    http_method_names = ["get", "post", "patch", "delete"]
+    queryset = Plan.objects.all()
+    # serializer_class = PlanSerializer
+
+    def get_serializer_class(self):
```
Comment (RahulRimal) [inline subscription/views.py:16]: remove this

File: subscription/views.py [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+        if self.request.method == "POST":
+            return CreatePlanSerializer
+        if self.request.method == "PATCH":
+            return UpdatePlanSerializer
+        return PlanSerializer
+
+class PlanCostViewSet(BaseApiMixin, ModelViewSet):
+    http_method_names = ["get", "post", "patch", "delete"]
+    queryset = PlanCost.objects.all()
+    # serializer_class = PlanCostSerializer
```
Comment (RahulRimal) [inline subscription/views.py:26]: uncomment this

File: subscription/views.py [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+
+    def get_serializer_class(self):
```
Comment (RahulRimal) [inline subscription/views.py:28]: remove this

File: ccib/settings.py [hunk 4]
```diff
@@ -25,7 +25,11 @@
 # SECURITY WARNING: don't run with debug turned on in production!
 DEBUG = True
 
-ALLOWED_HOSTS = []
+ALLOWED_HOSTS = [
```
[the remark this answers is not in the document]
Comment (arj9912) [reply ccib/settings.py:28]: ok arko localsetting banayarw rakhne ho

File: ccib/urls.py [hunk 5]
```diff
@@ -14,13 +14,17 @@
     1. Import the include() function: from django.urls import include, path
     2. Add a URL to urlpatterns:  path('blog/', include('blog.urls'))
 """
+import subscriptions
```
[the remark this answers is not in the document]
Comment (arj9912) [reply ccib/urls.py]: ok done

File: cooperative/serializers.py [hunk 6]
```diff
@@ -304,3 +308,19 @@ class Meta:
...
             "lone_taker_type",
         ]
+
+
+class FinaceSerializer(BaseModelSerializerMixin):
+    class Meta:
+        model = Finance
+        fields = ["idx", "name", "description", "location"]
+
+class CreateFinanceSerializer(BaseModelSerializerMixin):
```
[the remark this answers is not in the document]
Comment (arj9912) [reply cooperative/serializers.py]: ok done

File: cooperative/serializers.py [hunk 7]
```diff
@@ -304,3 +308,19 @@ class Meta:
...
+    class Meta:
+        model = Finance
+        fields = ["idx", "name", "description", "location"] 
+
+class UpdateFinanceSerializer(BaseModelSerializerMixin):
```
[the remark this answers is not in the document]
Comment (arj9912) [reply cooperative/serializers.py]: ok done

File: cooperative/views.py [hunk 8]
```diff
@@ -69,3 +70,16 @@ def get_serializer_class(self):
...
             return UpdateCompanySerializer
         return CompanySerializer
+
+
+class FinaceViewSet(BaseApiMixin, ModelViewSet):
+    http_method_names = ["get", "post", "patch", "delete"]
+    queryset = Finance.objects.all()
+    # serializer_class = FinaceSerializer
+
+    def get_serializer_class(self):
```
[the remark this answers is not in the document]
Comment (arj9912) [reply cooperative/views.py]: ok done

File: subscription/models.py [hunk 9]
```diff
@@ -0,0 +1,46 @@
+from django.db import models
+from common.models import BaseModelMixin
+from django_flex_subscriptions.models import SubscriptionBase, PlanBase
+from cooperative.models import Finance
+
+class Plan(BaseModelMixin, PlanBase):
+    PERIOD_MONTHLY = "monthly"
+    PERIOD_YEARLY = "yearly"
```
[the remark this answers is not in the document]
Comment (arj9912) [reply subscription/models.py:9]: ok done

File: subscription/models.py [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+        (STATUS_ACTIVE, STATUS_ACTIVE.capitalize()),
+        (STATUS_CANCELLED, STATUS_CANCELLED.capitalize()),
+        (STATUS_DUE, STATUS_DUE.capitalize()),
+    ]
+    
+    finance = models.ForeignKey(Finance, on_delete=models.CASCADE)
+    plan = models.ForeignKey(Plan, on_delete=models.CASCADE)
+    billing_start = models.DateField()
+    billing_end = models.DateField()
+    last_bill_paid = models.DateField()
```
[the remark this answers is not in the document]
Comment (arj9912) [reply subscription/models.py]: ok done

File: subscription/models.py [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+    next_billing = models.DateField()
```
[the remark this answers is not in the document]
Comment (arj9912) [reply subscription/models.py]: ok done

File: subscription/models.py [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+    grace_period = models.PositiveIntegerField(default=0)
+    status = models.CharField(max_length=10, choices=STATUS_CHOICES, default=STATUS_DUE)
+    verified = models.BooleanField(default=False)
```
[the remark this answers is not in the document]
Comment (arj9912) [reply subscription/models.py]: ok done

File: subscription/serializers.py [hunk 13]
```diff
@@ -0,0 +1,52 @@
+from common.mixins import BaseModelSerializerMixin
+from subscription.models import Plan, PlanCost, Subscription
+
+class PlanSerializer(BaseModelSerializerMixin):
+    class Meta:
+        model = Plan
+        fields = ["idx", "name", "description", "period"]
+
+class CreatePlanSerializer(BaseModelSerializerMixin):
```
[the remark this answers is not in the document]
Comment (arj9912) [reply subscription/serializers.py]: ok done

File: subscription/serializers.py [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+    class Meta:
+        model = Plan
+        fields = ["idx", "name", "description", "period"]
+
+class UpdatePlanSerializer(BaseModelSerializerMixin):
```
[the remark this answers is not in the document]
Comment (arj9912) [reply subscription/serializers.py]: ok done

File: subscription/serializers.py [hunk 15]
```diff
@@ -0,0 +1,52 @@
...
+        fields = ["idx", "name", "description", "period"]
+
+
+class PlanCostSerializer(BaseModelSerializerMixin):
+    class Meta:
+        model = PlanCost
+        fields = ["idx", "plan", "cost", "recurrance"]
+
+
+class CreatePlanCostSerializer(BaseModelSerializerMixin):
```
[the remark this answers is not in the document]
Comment (arj9912) [reply subscription/serializers.py]: ok done

File: subscription/serializers.py [hunk 16]
```diff
@@ -0,0 +1,52 @@
...
+    class Meta:
+        model = PlanCost
+        fields = ["idx", "plan", "cost", "recurrance"]
+    
+
+class UpdatePlanCostSerializer(BaseModelSerializerMixin):
```
[the remark this answers is not in the document]
Comment (arj9912) [reply subscription/serializers.py]: ok done

File: subscription/views.py [hunk 17]
```diff
@@ -0,0 +1,38 @@
+from django.shortcuts import render
+from rest_framework.viewsets import ModelViewSet
+
+from common.mixins import BaseApiMixin
+from subscription.models import Plan, PlanCost, Subscription
+from subscription.serializers import CreatePlanCostSerializer, CreatePlanSerializer, PlanCostSerializer, PlanSerializer, SubscriptionSerializer, UpdatePlanCostSerializer, UpdatePlanSerializer
+
+# Create your views here.
```
[the remark this answers is not in the document]
Comment (arj9912) [reply subscription/views.py]: ok done

File: subscription/views.py [hunk 18]
```diff
@@ -0,0 +1,38 @@
...
+
+
+class PlanViewSet(BaseApiMixin, ModelViewSet):
+    http_method_names = ["get", "post", "patch", "delete"]
+    queryset = Plan.objects.all()
+    # serializer_class = PlanSerializer
```
[the remark this answers is not in the document]
Comment (arj9912) [reply subscription/views.py]: ok done

File: subscription/views.py (hunk 1 above)
Comment (arj9912) [reply subscription/views.py]: ok done

File: subscription/views.py (hunk 2 above)
Comment (arj9912) [reply subscription/views.py]: ok done

File: subscription/views.py (hunk 3 above)
Comment (arj9912) [reply subscription/views.py]: ok done

File: ccib/settings.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (RahulRimal) [reply ccib/settings.py:28]: yes. BM hernu na testae ho

File: subscription/models.py [hunk 19]
```diff
@@ -33,14 +38,14 @@ class Subscription(BaseModelMixin, SubscriptionBase):
...
     ]
-    
+
     finance = models.ForeignKey(Finance, on_delete=models.CASCADE)
     plan = models.ForeignKey(Plan, on_delete=models.CASCADE)
-    billing_start = models.DateField()
-    billing_end = models.DateField()
-    last_bill_paid = models.DateField()
-    next_billing = models.DateField()
+    billing_start = models.DateField(null=True, blank=True)
```
Comment (RahulRimal) [inline subscription/models.py:44]: yo null and blank hudaina

File: subscription/models.py [hunk 20]
```diff
@@ -33,14 +38,14 @@ class Subscription(BaseModelMixin, SubscriptionBase):
...
+    billing_end = models.DateField(null=True, blank=True)
```
Comment (RahulRimal) [inline subscription/models.py:45]: yo null and blank hudaina

File: subscription/models.py [hunk 21]
```diff
@@ -33,14 +38,14 @@ class Subscription(BaseModelMixin, SubscriptionBase):
...
+    last_bill_paid = models.DateField(null=True, blank=True)
+    next_billing = models.DateField(null=True, blank=True)
```
Comment (RahulRimal) [inline subscription/models.py:47]: yo null and blank hudaina

**Review by RahulRimal — CHANGES_REQUESTED** [review]

File: ccib/settings.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (arj9912) [reply ccib/settings.py:28]: eaa lwlw hunxa

File: subscription/models.py (hunk 21 above)
Comment (arj9912) [reply subscription/models.py]: ok done

File: subscription/models.py (hunk 20 above)
Comment (arj9912) [reply subscription/models.py]: ok done

File: ccib/settings.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (arj9912) [reply ccib/settings.py:28]: ok dai

File: ccib/urls.py [hunk 22]
```diff
@@ -15,12 +15,14 @@
...
 from django.contrib import admin
-from django.urls import path, include
+from django.urls import path, include 
 
 urlpatterns = [
+    path('subscriptions/', include('subscriptions.urls')),
     path('admin/', admin.site.urls),
     path('auth/', include('autho.urls')),
     path('cooperative/', include('cooperative.urls')),
+    path('subscription/', include('subscription.urls')),
```
Comment (RahulRimal) [inline ccib/urls.py:25]: remove this

**Review by RahulRimal — CHANGES_REQUESTED** [review]

File: ccib/urls.py (hunk 22 above)
Comment (arj9912) [reply ccib/urls.py:24]: ok done

**Review by RahulRimal — APPROVED** [review]

**Pull request merged**